blaze/expr/collections.py

```diff
@@ -437,6 +440,102 @@ def join(lhs, rhs, on_left=None, on_right=None,
 join.__doc__ = Join.__doc__
 
 
+class Concat(Expr):
+
+    """ Stack tables on common columns
+
+    Parameters
+    ----------
+    lhs : Expr
+    rhs : Expr
+    axis : int, optional
+        The axis to concatenate on.
+
+    Examples
+    --------
+    >>> from blaze import symbol
+    >>> names = symbol('names', '5 * {name: string, id: int32}')
+    >>> more_names = symbol('more_names', '7 * {name: string, id: int32}')
+
+    Vertically stack these tables.
+    >>> stacked = concat(names, more_names)
+    >>> stacked.dshape
+    dshape("12 * {name: string, id: int32}")
+
+    >>> mat_a = symbol('a', '3 * 5 * int32')
+    >>> mat_b = symbol('b', '3 * 5 * int32')
+
+    Vertically stack these matricies.
+    >>> vstacked = concat(mat_a, mat_b, axis=0)
+    >>> vstacked.dshape
+    dshape("6 * 5 * int32")
+
+    Horizontally stack these matricies.
+    >>> hstacked = concat(mat_a, mat_b, axis=1)
+    >>> hstacked.dshape
+    dshape("3 * 10 * int32")
+
+    See Also
+    --------
+
+    blaze.expr.collections.Merge
+    """
+    __slots__ = '_hash', 'lhs', 'rhs', 'axis'
+    __inputs__ = 'lhs', 'rhs'
+
+    @property
+    def dshape(self):
+        axis = self.axis
+        ldshape = self.lhs.dshape
+        lshape = ldshape.shape
+        return DataShape(
+            *(lshape[:axis] + (
+                _shape_add(lshape[axis], self.rhs.dshape.shape[axis]),
+            ) + lshape[axis + 1:] + (ldshape.measure,))
+        )
+
+
+def _shape_add(a, b):
+    if isinstance(a, Var) or isinstance(b, Var):
+        return var
+    return Fixed(a.val + b.val)
+
+
+def concat(lhs, rhs, axis=0):
+    if axis < 0:
+        raise ValueError('axis must be greater than or equal to 0')
+
+    ldshape = lhs.dshape
+    rdshape = rhs.dshape
+    if ldshape.measure != rdshape.measure:
+        raise ValueError(
+            'Mismatched measures: {l} != {r}'.format(
+                l=ldshape.measure, r=rdshape.measure
+            ),
+        )
+
+    lshape = ldshape.shape
+    rshape = rdshape.shape
+    for n, (a, b) in enumerate(zip_longest(lshape, rshape, fillvalue=None)):
+        if n != axis and a != b:
+            raise ValueError(
+                "Shapes are not equal along axis '{n}': {a} != {b}".format(
```

Comment:
I will drop them. Do you think this should also be a type error since it is about the shape of the expr?